File: binding-java/runtime/src/test/java/org/apache/etch/bindings/java/support/TestValidator_boolean.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.bindings.java.support;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

import java.util.Date;

import org.apache.etch.bindings.java.msg.StructValue;
import org.apache.etch.bindings.java.msg.Type;
import org.apache.etch.bindings.java.msg.Validator;
import org.apache.etch.bindings.java.msg.ValueFactory;
import org.junit.Test;


/**
 * Test Validator_boolean.
 */
public class TestValidator_boolean
{
	private final ValueFactory vf = new DummyValueFactory();
	
	/** @throws Exception */
	@Test
	public void constructor1() throws Exception
	{
		testconstructor( 0, "boolean[0]", Boolean.class );
		testconstructor( 1, "boolean[1]", boolean[].class );
		testconstructor( 2, "boolean[2]", boolean[][].class );
		testconstructor( 3, "boolean[3]", boolean[][][].class );
		testconstructor( 4, "boolean[4]", boolean[][][][].class );
		testconstructor( 5, "boolean[5]", boolean[][][][][].class );
		testconstructor( 6, "boolean[6]", boolean[][][][][][].class );
		testconstructor( 7, "boolean[7]", boolean[][][][][][][].class );
		testconstructor( 8, "boolean[8]", boolean[][][][][][][][].class );
		testconstructor( 9, "boolean[9]", boolean[][][][][][][][][].class );
		assertEquals( 9, Validator.MAX_NDIMS );
	}
	
	private void testconstructor( int n, String descr, Class<?> expectedClass )
	{
		Validator_boolean v = Validator_boolean.get( n );
		assertEquals( n, v.getNDims() );
		assertSame( expectedClass, v.getExpectedClass() );
		assertEquals( descr, v.toString() );
	}
	
	/** @throws Exception */
	@Test( expected = IllegalArgumentException.class )
	public void constructor2() throws Exception
	{
		Validator_boolean.get( -1 );
	}
	
	/** @throws Exception */
	@Test( expected = IllegalArgumentException.class )
	public void constructor3() throws Exception
	{
		Validator_boolean.get( Validator.MAX_NDIMS+1 );
	}
	
	/** @throws Exception */
	@Test
	public void elementvalidator1() throws Exception
	{
		testelementvalidator( 1, "boolean[0]", Boolean.class );
		testelementvalidator( 2, "boolean[1]", boolean[].class );
		testelementvalidator( 3, "boolean[2]", boolean[][].class );
		testelementvalidator( 4, "boolean[3]", boolean[][][].class );
		testelementvalidator( 5, "boolean[4]", boolean[][][][].class );
		testelementvalidator( 6, "boolean[5]", boolean[][][][][].class );
		testelementvalidator( 7, "boolean[6]", boolean[][][][][][].class );
		testelementvalidator( 8, "boolean[7]", boolean[][][][][][][].class );
		testelementvalidator( 9, "boolean[8]", boolean[][][][][][][][].class );
		assertEquals( 9, Validator.MAX_NDIMS );
	}
	
	private void testelementvalidator( int n, String descr, Class<?> expectedClass )
	{
		Validator_boolean v = (Validator_boolean) Validator_boolean.get( n ).elementValidator();
		assertEquals( n-1, v.getNDims() );
		assertSame( expectedClass, v.getExpectedClass() );
		assertEquals( descr, v.toString() );
	}

	/** @throws Exception */
	@Test( expected = IllegalArgumentException.class )
	public void elementvalidator2() throws Exception
	{
		Validator_boolean.get( 0 ).elementValidator();
	}
	
	/** @throws Exception */
	@Test
	public void good_scalar() throws Exception
	{
		testgoodvalue( 0, false );
		testgoodvalue( 0, true );
	}
	
	/** @throws Exception */
	@Test
	public void good_array() throws Exception
	{
		testgoodvalue( 1, new boolean[] {} );
		testgoodvalue( 2, new boolean[][] {} );
		testgoodvalue( 3, new boolean[][][] {} );
		testgoodvalue( 4, new boolean[][][][] {} );
		testgoodvalue( 5, new boolean[][][][][] {} );
		testgoodvalue( 6, new boolean[][][][][][] {} );
		testgoodvalue( 7, new boolean[][][][][][][] {} );
		testgoodvalue( 8, new boolean[][][][][][][][] {} );
		testgoodvalue( 9, new boolean[][][][][][][][][] {} );
		assertEquals( 9, Validator.MAX_NDIMS );
	}
	
	private void testgoodvalue( int n, Object value )
	{
		TypeValidator v = Validator_boolean.get( n );
		assertTrue( v.validate( value ) );
		assertTrue( validateValueOk( v, value ) );
	}
	
	/** @throws Exception */
	@Test
	public void bad_scalar() throws Exception
	{
		testbadvalue( 0, null );
		testbadvalue( 0, (byte) 1 );
		testbadvalue( 0, (short) 2222 );
		testbadvalue( 0, 33333333 );
		testbadvalue( 0, 4444444444444444L );
		testbadvalue( 0, 5.5f );
		testbadvalue( 0, 6.6 );
		testbadvalue( 0, "" );
		testbadvalue( 0, "abc" );
		testbadvalue( 0, new Object() );
		testbadvalue( 0, new StructValue( new Type( "abc" ), vf ) );
		testbadvalue( 0, new Date() );

		testbadvalue( 1, null );
		testbadvalue( 1, false );
		testbadvalue( 1, true );
		testbadvalue( 1, (byte) 1 );
		testbadvalue( 1, (short) 2222 );
		testbadvalue( 1, 33333333 );
		testbadvalue( 1, 4444444444444444L );
		testbadvalue( 1, 5.5f );
		testbadvalue( 1, 6.6 );
		testbadvalue( 1, "" );
		testbadvalue( 1, "abc" );
		testbadvalue( 1, new Object() );
		testbadvalue( 1, new StructValue( new Type( "abc" ), vf ) );
		testbadvalue( 1, new Date() );
	}
	
	/** @throws Exception */
	@Test
	public void bad_array() throws Exception
	{
		testbadvalue( 0, new boolean[] {} );
		testbadvalue( 1, new boolean[][] {} );
		testbadvalue( 2, new boolean[][][] {} );
		testbadvalue( 3, new boolean[][][][] {} );
		testbadvalue( 4, new boolean[][][][][] {} );
		testbadvalue( 5, new boolean[][][][][][] {} );
		testbadvalue( 6, new boolean[][][][][][][] {} );
		testbadvalue( 7, new boolean[][][][][][][][] {} );
		testbadvalue( 8, new boolean[][][][][][][][][] {} );
		testbadvalue( 9, new boolean[][][][][][][][][][] {} );
		assertEquals( 9, Validator.MAX_NDIMS );

		testbadvalue( 2, new boolean[] {} );
		testbadvalue( 3, new boolean[][] {} );
		testbadvalue( 4, new boolean[][][] {} );
		testbadvalue( 5, new boolean[][][][] {} );
		testbadvalue( 6, new boolean[][][][][] {} );
		testbadvalue( 7, new boolean[][][][][][] {} );
		testbadvalue( 8, new boolean[][][][][][][] {} );
		testbadvalue( 9, new boolean[][][][][][][][] {} );
		assertEquals( 9, Validator.MAX_NDIMS );
	}

	private void testbadvalue( int n, Object value )
	{
		TypeValidator v = Validator_boolean.get( n );
		assertFalse( v.validate( value ) );
		assertFalse( validateValueOk( v, value ) );
	}

	private boolean validateValueOk( Validator v, Object value )
	{
		try
		{
			Object x = v.validateValue( value );
			assertEquals( value, x );
			return true;
		}
		catch ( Exception e )
		{
			return false;
		}
	}
}

File: binding-csharp/runtime/src/test/csharp/Etch/Support/TestValidator_boolean.cs
// $Id$
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.
// 
using System;
using Org.Apache.Etch.Bindings.Csharp.Msg;
using NUnit.Framework;

namespace Org.Apache.Etch.Bindings.Csharp.Support
{
    [TestFixture]
    public class TestValidator_Boolean
    {
        private readonly ValueFactory vf = new DummyValueFactory();

        [TestFixtureSetUp]
        public void First()
        {
            Console.WriteLine();
            Console.Write( "TestValidator_Boolean" );
        }

        [Test]
	    public void Constructor1()  
	    {
		    TestConstructor(0,"boolean[0]",typeof(Boolean));
            TestConstructor(1, "boolean[1]", typeof(Boolean[]));
            TestConstructor(2, "boolean[2]", typeof(Boolean[][]));
            TestConstructor(3, "boolean[3]", typeof(Boolean[][][]));
            TestConstructor(4, "boolean[4]", typeof(Boolean[][][][]));
            TestConstructor(5, "boolean[5]", typeof(Boolean[][][][][]));
            TestConstructor(6, "boolean[6]", typeof(Boolean[][][][][][]));
            TestConstructor(7, "boolean[7]", typeof(Boolean[][][][][][][]));
            TestConstructor(8, "boolean[8]", typeof(Boolean[][][][][][][][]));
            TestConstructor(9, "boolean[9]", typeof(Boolean[][][][][][][][][]));
            Assert.AreEqual(9,Validator.MAX_NDIMS);
	    }

        private void TestConstructor(int n, string descr, Type expectedClass)
        {
            Validator_boolean v = Validator_boolean.Get(n);
            Assert.AreEqual(n, v.GetNDims());
            Assert.AreSame(expectedClass,v.GetExpectedClass());
            Assert.AreEqual(descr, v.ToString());            
        }
 
	    [Test]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
	    public void Constructor2()
	    {		
		    Validator_boolean.Get( -1 );
	    }

        [Test]
        [ExpectedException(typeof(ArgumentOutOfRangeException))]
        public void Constructor3()
        {
            Validator_boolean.Get(Validator.MAX_NDIMS + 1);
        }

        [Test]
        public void ElementValidator1() 
	    {
		    TestElementValidator( 1, "boolean[0]", typeof(Boolean)  );
		    TestElementValidator( 2, "boolean[1]", typeof(Boolean[])  );
		    TestElementValidator( 3, "boolean[2]", typeof(Boolean[][])  );
		    TestElementValidator( 4, "boolean[3]", typeof(Boolean[][][])  );
		    TestElementValidator( 5, "boolean[4]", typeof(Boolean[][][][])  );
		    TestElementValidator( 6, "boolean[5]", typeof(Boolean[][][][][])  );
		    TestElementValidator( 7, "boolean[6]", typeof(Boolean[][][][][][])  );
		    TestElementValidator( 8, "boolean[7]", typeof(Boolean[][][][][][][])  );
		    TestElementValidator( 9, "boolean[8]", typeof(Boolean[][][][][][][][])  );
		    Assert.AreEqual( 9, Validator.MAX_NDIMS );
	    }

        private void TestElementValidator( int n, String descr, Type expectedClass )
	    {
		    Validator_boolean v = (Validator_boolean) Validator_boolean.Get( n ).ElementValidator();
		    Assert.AreEqual( n-1, v.GetNDims() );
		    Assert.AreSame( expectedClass, v.GetExpectedClass() );
		    Assert.AreEqual( descr, v.ToString() );
	    }


        [Test]
        [ExpectedException( typeof( ArgumentOutOfRangeException ) )]
	    public void ElementValidator2  ()
	    {		
		    Validator_boolean.Get( 0 ).ElementValidator();
        }

       
	    [Test]
	    public void Good_Scalar() 
	    {
		    TestGoodValue( 0,  false );
		    TestGoodValue( 0, true );
	    }
	
	
	    [Test]
	    public void Good_Array() 
	    {
		    TestGoodValue( 1, new Boolean[] {} );
		    TestGoodValue( 2,  new Boolean[][] {} );
		    TestGoodValue( 3,  new Boolean[][][] {} );
		    TestGoodValue( 4,  new Boolean[][][][] {} );
		    TestGoodValue( 5,  new Boolean[][][][][] {} );
		    TestGoodValue( 6,  new Boolean[][][][][][] {} );
		    TestGoodValue( 7,  new Boolean[][][][][][][] {} );
		    TestGoodValue( 8,  new Boolean[][][][][][][][] {} );
		    TestGoodValue( 9,  new Boolean[][][][][][][][][] {} );
		    Assert.AreEqual( 9, Validator.MAX_NDIMS );
	    }
	
	    private void TestGoodValue( int n,  Object value )
	    {
		    TypeValidator v = Validator_boolean.Get( n );
		   
		    Assert.IsTrue( v.Validate( value ) );
		    Assert.IsTrue( ValidateValueOk( v, value ) );
	    }
	
	
	    [Test]
	    public void bad_scalar() 
	    {
		    TestBadValue( 0, null );
		    TestBadValue( 0, (byte) 1 );
		    TestBadValue( 0, (short) 2222 );
		    TestBadValue( 0, 33333333 );
		    TestBadValue( 0, 4444444444444444L );
		    TestBadValue( 0, 5.5f );
		    TestBadValue( 0, 6.6 );
		    TestBadValue( 0, "" );
		    TestBadValue( 0, "abc" );
		    TestBadValue( 0, new Object() );
            TestBadValue(0, new StructValue(new XType("abc"), vf));
		    TestBadValue( 0, new DateTime() );

		    TestBadValue( 1, null );
		    TestBadValue( 1, false );
		    TestBadValue( 1, true );
		    TestBadValue( 1, (byte) 1 );
		    TestBadValue( 1, (short) 2222 );
		    TestBadValue( 1, 33333333 );
		    TestBadValue( 1, 4444444444444444L );
		    TestBadValue( 1, 5.5f );
		    TestBadValue( 1, 6.6 );
		    TestBadValue( 1, "" );
		    TestBadValue( 1, "abc" );
		    TestBadValue( 1, new Object() );
            TestBadValue(1, new StructValue(new XType("abc"), vf));
		    TestBadValue( 1, new DateTime() );
	    }
	
	
	    [Test]
	    public void bad_array() 
	    {
		    TestBadValue( 0, new Boolean[] {} );
		    TestBadValue( 1, new Boolean[][] {} );
		    TestBadValue( 2, new Boolean[][][] {} );
		    TestBadValue( 3, new Boolean[][][][] {} );
		    TestBadValue( 4, new Boolean[][][][][] {} );
		    TestBadValue( 5, new Boolean[][][][][][] {} );
		    TestBadValue( 6, new Boolean[][][][][][][] {} );
		    TestBadValue( 7, new Boolean[][][][][][][][] {} );
		    TestBadValue( 8, new Boolean[][][][][][][][][] {} );
		    TestBadValue( 9, new Boolean[][][][][][][][][][] {} );
		    Assert.AreEqual( 9, Validator.MAX_NDIMS );

		    TestBadValue( 2, new Boolean[] {} );
		    TestBadValue( 3, new Boolean[][] {} );
		    TestBadValue( 4, new Boolean[][][] {} );
		    TestBadValue( 5, new Boolean[][][][] {} );
		    TestBadValue( 6, new Boolean[][][][][] {} );
		    TestBadValue( 7, new Boolean[][][][][][] {} );
		    TestBadValue( 8, new Boolean[][][][][][][] {} );
		    TestBadValue( 9, new Boolean[][][][][][][][] {} );
            Assert.AreEqual(9, Validator.MAX_NDIMS);
	    }

	    private void TestBadValue( int n, Object value )
	    {
		    TypeValidator v = Validator_boolean.Get( n );
		    Assert.IsFalse( v.Validate( value ) );
		    Assert.IsFalse( ValidateValueOk( v, value ) );
	    }

	    private bool ValidateValueOk( Validator v, Object value )
	    {
		    try
		    {
			    Object x = v.ValidateValue( value );
			    Assert.AreEqual( value, x );
			    return true;
		    }
		    catch ( Exception )
		    {
               
                return false;
		    }
	    }




    }
}
